controllers/loadtest_controller.go

```diff
@@ -56,64 +57,44 @@ type LoadTestReconciler struct {
 // or handling the termination of its pods.
 func (r *LoadTestReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	var err error
-
 	log := r.Log.WithValues("loadtest", req.NamespacedName)
+
 	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
 	defer cancel()
 
-	// Fetch the current state of the world.
-
-	// Listing nodes will be required to achieve gang scheduling; however, this is
-	// not implemented at this time. Therefore, it is commented here for later:
-	//
-	//nodes := new(corev1.NodeList)
-	//if err = r.List(ctx, nodes); err != nil {
-	//	log.Error(err, "failed to list nodes")
-	//	return ctrl.Result{Requeue: true}, err
-	//}
-
-	pods := new(corev1.PodList)
-	if err = r.List(ctx, pods, client.InNamespace(req.Namespace)); err != nil {
-		log.Error(err, "failed to list pods", "namespace", req.Namespace)
-		return ctrl.Result{Requeue: true}, err
-	}
-
-	loadtests := new(grpcv1.LoadTestList)
-	if err = r.List(ctx, loadtests); err != nil {
-		log.Error(err, "failed to list loadtests")
-		return ctrl.Result{Requeue: true}, err
-	}
-
 	loadtest := new(grpcv1.LoadTest)
 	if err = r.Get(ctx, req.NamespacedName, loadtest); err != nil {
 		log.Error(err, "failed to get loadtest", "name", req.NamespacedName)
-
-		// do not requeue, the load test may have been deleted
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
 	if loadtest.Status.State.IsTerminated() {
-		// the load test has already completed, this is probably garbage collection
 		return ctrl.Result{}, nil
 	}
 
-	loadtest = loadtest.DeepCopy()
-	if err = r.Defaults.SetLoadTestDefaults(loadtest); err != nil {
-		log.Error(err, "failed to set defaults on loadtest")
-
-		// do not requeue, something has gone horribly wrong
+	// TODO(codeblooded): Consider moving this to a mutating webhook
+	loadtestWithDefaults := loadtest.DeepCopy()
+	if err = r.Defaults.SetLoadTestDefaults(loadtestWithDefaults); err != nil {
+		log.Error(err, "failed to clone test with defaults")
 		return ctrl.Result{}, err
 	}
-	if err = r.Update(ctx, loadtest); err != nil {
-		log.Error(err, "failed to update loadtest with defaults")
+	if !reflect.DeepEqual(loadtest, loadtestWithDefaults) {
+		if err = r.Update(ctx, loadtestWithDefaults); err != nil {
+			log.Error(err, "failed to update test with defaults")
+			return ctrl.Result{}, err
```

Comment:
Little confused about the update function, I used to think it happens like this: every time we get a triggering LoadTest -> we made sure that it follows the default -> we update the triggering LoadTest. I understand you make the change to make it a lazy update, but say if we do found a difference (that is line 82), why are we updating the `loadtestWithDefaults` instead of `loadtest`?

Reply:
`loadtestWithDefaults` is equivalent to `loadtest`, except it adds defaults. It's name, namespace, and GVK (group, version, kind) information is identical. If it looks like a duck... walks like a duck... quacks like a duck... it is a duck. To Kubernetes, this is enough for it to know which object we are talking about (even if it  is not the same object in memory). It will be treated the same. We don't want to update with loadtest, because that would do nothing. It has no defaults. They both are kept around, b/c line 81 ensures that we do not update every single time we reconcile (only when we need to set defaults).

Reply:
" To Kubernetes, this is enough for it to know which object we are talking about (even if it is not the same object in memory)" Got it ! Thx. :)